Folio Driver | getMyTransactions | add correctly formatted dueDate and dueStatus

cwolfebsco opened the pull request:
Current State:
Folio Driver would not indicate whether an loaned item (from /circulation/loans endpoint) was due or overdue

New State:
Folio Driver will:
* consider an item due if it's dueDate timestamp is within 24 hours of now
* consider an item overdue, if now is greater than dueTime
* set false (neither due nor overdue) in all other cases

This will display bootstrap alert divs next to relevant loans, additionally it will update the pills in the my account menu appropriately

cwolfebsco committed:
Feature: adding timezone adjusted dueDate, dueTime and dueStatus

dueDate & dueTime are adjusted for Timezone. Folio responds in UTC, but your local timezone might be earlier / later causing unpleasant display issues.

dueStatus is:
* overdue if the current unix timestamp is greater than the dueDate timestamp
* due if the current unix timestamp is within 24 hours
* false in all other cases

cwolfebsco committed:
adding namespaces and fixing whitespaces

added required namespaces
fixed whitespace requirements

demiankatz commented on module/VuFind/src/VuFind/ILS/Driver/Folio.php:
```diff
@@ -897,12 +899,23 @@ public function getMyTransactions($patron)
+            $date->setTimezone($loc);
+
+            $dueStatus = false;
+            $tmpDueDate = strtotime($trans->dueDate);
```
Should you derive $tmpDueDate from $date above, instead of going back to the raw $trans->dueDate? Just wondering if this impacts the time zone handling.

cwolfebsco replied:
Another good question :) I was under the impression that the strtotime of a ISO8601 string (e.g. 2022-06-21T10:32:07.860+00:00) would always result in the correct Unix timestamp as the timezone was indicated at the end. On the other hand there would be no harm done by using `$tmpDueDate = $date->getTimestamp()`

I've had this snippet run in Trinity's VuFind for a while (but I must also admit that it originated from a much older version of the Driver). one to ponder for sure

demiankatz replied:
You are probably correct that it will work either way, but I suspect switching it to use `$date->getTimestamp()` would be more efficient since it would eliminate the need to parse the date twice, which I'm sure is a more expensive operation than fetching an already-calculated timestamp.

cwolfebsco replied:
I like the efficiency argument and I am rather embarrassed that I didn't think about that :) Thanks for pointing it out!

demiankatz reviewed (commented):
Thanks for this, @cwolfebsco -- just one question!

cwolfebsco committed:
Improve efficiency of tmpDueDate calculation

Based on Demian's feedback, using date->getTimestamp() for efficiency, avoiding having to parse the string twice. thanks for the feedback Demian

demiankatz commented on module/VuFind/src/VuFind/ILS/Driver/Folio.php:
```diff
@@ -897,12 +899,23 @@ public function getMyTransactions($patron)
+            $date->setTimezone($loc);
+
+            $dueStatus = false;
+            $tmpDueDate = $date->getTimestamp();
```
Maybe `$dueDateTimestamp` would be a more clear name for this variable.

demiankatz commented on module/VuFind/src/VuFind/ILS/Driver/Folio.php:
```diff
@@ -897,12 +899,23 @@ public function getMyTransactions($patron)
-                'dueTime' => date_format($date, "g:i:s a"),
-                // TODO: Due Status
-                // 'dueStatus' => $trans['itemId'],
+                'duedate' => $date->format('d F Y'),
```
Maybe we could here use `$this->dateConverter->convertToDisplayDate('U', $tmpDueDate)` and `$this->dateConverter->convertToDisplayTime('U', $tmpDueDate)` to take advantage of configurable date/time formats, similar to what we discussed on #2484. (And of course, please replace $tmpDueDate with the appropriate value if you choose to rename it per my suggestion above).

demiankatz commented on module/VuFind/src/VuFind/ILS/Driver/Folio.php:
```diff
@@ -897,12 +899,23 @@ public function getMyTransactions($patron)
         ) as $trans) {
-            $date = date_create($trans->dueDate);
+            $date = new DateTime($trans->dueDate, new DateTimeZone('UTC'));
+            $loc = (new DateTime)->getTimezone();
```
Maybe `$localTimezone` instead of just `$loc` for greater clarity?

demiankatz reviewed (changes requested):
Thanks for the progress on this. A few more very minor suggestions for polish, if you don't mind!

cwolfebsco committed:
Improve clarity of variable naming & implement consistent date / time formatting

Implementing Demian's suggestions, which improve clarity of variable naming
$loc => $localTimezone
$tmpDueDate => $duedateTimestamp

Additionally implemented $this->dateConverter->convertToDisplaydate for improved consistency of date / time formatting within the driver

Additional changes were introduced for code format requirements (not sure those are best choices)

demiankatz commented:
Thanks, @cwolfebsco, this looks good! I'm going to do a little testing later today to make sure this works as expected in my test environment, and if all goes well, I will merge it.

demiankatz reviewed (approved):
I've confirmed that this corrects inaccurate dates in my test environment. Thanks again, @cwolfebsco!

demiankatz commented:
Since this is a bug fix, I've also backported it to the release-8.1 branch.
